src/postgres-util/src/lib.rs

```diff
@@ -309,50 +317,88 @@ impl Config {
+                let tcp_stream = TokioTcpStream::connect(("localhost", local_port)).await?;
                 let (client, connection) = self
                     .postgres_config
                     .replication_mode(ReplicationMode::Logical)
                     .connect_timeout(Duration::from_secs(30))
                     .keepalives_idle(Duration::from_secs(10 * 60))
-                    .connect_raw(channel, tls)
+                    .connect_raw(tcp_stream, tls)
                     .await?;
-                task::spawn(|| "postgres_connect_replication", connection);
+                task::spawn(|| "postgres_connect_replication", async {
+                    _ = connection.await;
+                    _ = session.close().await;
+                });
                 Ok(client)
             }
         }
     }
 
-    async fn establish_ssh_connection(
-        &self,
-    ) -> Result<AsyncChannel<TokioTcpStream>, anyhow::Error> {
+    async fn establish_openssh_connection(&self) -> Result<(openssh::Session, u16), anyhow::Error> {
         match &self.ssh_tunnel {
             SshTunnelConfig::Tunnel {
                 host,
                 port,
                 user,
-                public_key,
+                public_key: _,
                 private_key,
             } => {
-                let tcp_stream = TokioTcpStream::connect((host.as_str(), *port)).await?;
-                let mut session = AsyncSession::new(tcp_stream, Some(SessionConfiguration::new()))?;
-                session.handshake().await?;
+                let tempdir = tempfile::Builder::new()
+                    .prefix("ssh-tunnel-key")
+                    .tempdir()?;
+                let path = tempdir.path().join("key");
+                let mut tempfile = File::create(&path)?;
+                tempfile.set_permissions(std::fs::Permissions::from_mode(0o600))?;
+                tempfile.write_all(private_key.as_bytes())?;
+                tempfile.set_permissions(std::fs::Permissions::from_mode(0o400))?;
 
-                session
-                    .userauth_pubkey_memory(user, Some(public_key), private_key, None)
-                    .await?;
+                let mut builder = openssh::SessionBuilder::default();
+                builder.user(user.clone()).port(*port).keyfile(&path);
+                let session = builder.connect(host).await?;
 
-                if !session.authenticated() {
-                    bail!("failed SSH authentication")
-                };
+                // Delete the private key for safety: since `ssh` still has an open handle to it,
+                // it still has access to the key.
+                drop(tempfile);
+                std::fs::remove_file(&path)?;
+                drop(tempdir);
 
-                // Advertise the source connection to Postgres as coming from the bastion host
-                let src = Some((host.as_str(), *port));
+                // Ensure session is healthy
+                session.check().await?;
```

Comment:
is this strictly required? just wondering!

Reply:
Not required, no! I think it was helpful while debugging and it just stayed there.